src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImpl.java

```diff
@@ -19,63 +27,186 @@
+        pullRequests.getOrDefault(cacheKey, new RepositoryStore(new ConcurrentHashMap<>())).closePullRequest(fromBranch, toBranch);
     }
 
     @Override
     public boolean hasOpenPullRequests(String branchName, BitbucketSCMRepository repository) {
-
         PullRequestStoreImpl.CacheKey key =
                 new PullRequestStoreImpl.CacheKey(repository.getProjectKey(), repository.getRepositorySlug(), repository.getServerId());
-        return pullRequests.getOrDefault(key, new ConcurrentLinkedQueue<>())
-                .stream()
-                .filter(pullRequest -> pullRequest.getFromRef().getDisplayId().equals(branchName))
-                .findFirst()
-                .isPresent();
+        return pullRequests.getOrDefault(key, new RepositoryStore(new ConcurrentHashMap<>())).hasOpenPullRequests(branchName);
     }
 
     @Override
-    public Optional<BitbucketPullRequest> getPullRequest(String key, String slug, String serverId, int pullRequestId) {
+    public Optional<MinimalPullRequest> getPullRequest(String projectKey, String slug, String serverId, String fromBranch, String toBranch) {
         PullRequestStoreImpl.CacheKey cacheKey =
-                new PullRequestStoreImpl.CacheKey(key, slug, serverId);
-        ConcurrentLinkedQueue<BitbucketPullRequest> pullRequest = pullRequests.get(cacheKey);
-        if (pullRequest == null || pullRequest.isEmpty()) {
-            return Optional.empty();
-        }
-        return pullRequest.stream().filter(pr -> pr.getId() == pullRequestId).findFirst();
+                new PullRequestStoreImpl.CacheKey(projectKey, slug, serverId);
+        return Optional.ofNullable(pullRequests.getOrDefault(cacheKey, new RepositoryStore(new ConcurrentHashMap<>()))
+                .getPullRequest(fromBranch, toBranch));
     }
 
     @Override
-    public void refreshStore(String key, String slug, String serverId, Stream<BitbucketPullRequest> bbsPullRequests) {
+    public void refreshStore(String projectKey, String slug, String serverId, Stream<BitbucketPullRequest> bbsPullRequests) {
         PullRequestStoreImpl.CacheKey cacheKey =
-                new PullRequestStoreImpl.CacheKey(key, slug, serverId);
-        ConcurrentLinkedQueue<BitbucketPullRequest> pullRequest = pullRequests.get(cacheKey);
-        if (pullRequest != null) {
-            pullRequest.clear();
+                new PullRequestStoreImpl.CacheKey(projectKey, slug, serverId);
+        AtomicLong oldestUpdate = pullRequests.getOrDefault(cacheKey, new RepositoryStore(new ConcurrentHashMap<>())).findOldestUpdate();
+
+        //bbspullrequests is ordered from newest to oldest
+        bbsPullRequests.map(pr -> {
+            if (pr.getUpdatedDate() > oldestUpdate.get()) {
+                updatePullRequest(serverId, pr);
+            }
+            return pr.getUpdatedDate();
+        }).filter(updated -> updated < oldestUpdate.get())
+                .findFirst();
```

Comment:
Streams are lazy, so in order to make the stream 'do something' you need a terminal operation on it. `findFirst` means it will stop once the first item passes through the filter and we break the endless stream.

Reply:
Okay- it looks very strange to my eyes, so perhaps a comment to that effect would be good